lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1126,28 +1187,105 @@ class InlineExecuteRegionOpPattern
   }
 };
 
+static std::pair<SmallVector<calyx::PortInfo>, SmallVector<calyx::PortInfo>>
+getPortsForExternalMemref(PatternRewriter &rewriter, StringRef memName,
+                          Value memref) {
+  SmallVector<calyx::PortInfo> inPorts, outPorts;
+  MemRefType memrefType = memref.getType().cast<MemRefType>();
+
+  /// Read data
+  inPorts.push_back(
+      calyx::PortInfo{rewriter.getStringAttr(memName + "_read_data"),
+                      memrefType.getElementType(), calyx::Direction::Input,
+                      DictionaryAttr::get(rewriter.getContext(), {})});
+
+  /// Done
+  inPorts.push_back(calyx::PortInfo{
+      rewriter.getStringAttr(memName + "_done"), rewriter.getI1Type(),
+      calyx::Direction::Input, DictionaryAttr::get(rewriter.getContext(), {})});
+
+  /// Write data
+  outPorts.push_back(
+      calyx::PortInfo{rewriter.getStringAttr(memName + "_write_data"),
+                      memrefType.getElementType(), calyx::Direction::Output,
+                      DictionaryAttr::get(rewriter.getContext(), {})});
+
+  /// Memory address outputs
+  for (auto dim : enumerate(memrefType.getShape())) {
+    outPorts.push_back(calyx::PortInfo{
+        rewriter.getStringAttr(memName + "_addr" + std::to_string(dim.index())),
+        rewriter.getIntegerType(dim.value()), calyx::Direction::Output,
+        DictionaryAttr::get(rewriter.getContext(), {})});
+  }
+
+  /// Write enable
+  outPorts.push_back(
+      calyx::PortInfo{rewriter.getStringAttr(memName + "_write_en"),
+                      rewriter.getI1Type(), calyx::Direction::Output,
+                      DictionaryAttr::get(rewriter.getContext(), {})});
+
+  return {inPorts, outPorts};
+}
+
 /// Creates a new Calyx component for each FuncOp in the program.
 struct FuncOpConversion : public FuncOpPartialLoweringPattern {
   using FuncOpPartialLoweringPattern::FuncOpPartialLoweringPattern;
 
   LogicalResult
   PartiallyLowerFuncToComp(mlir::FuncOp funcOp,
                            PatternRewriter &rewriter) const override {
-    /// Create I/O ports.
-    SmallVector<calyx::PortInfo> ports;
+    /// Maintain a mapping between funcOp input arguments and the port index
+    /// which the argument will eventually map to.
+    DenseMap<Value, unsigned> funcOpArgRewrites;
+
+    /// Maintain a mapping between funcOp output indexes and the component
+    /// output port index which the return value will eventually map to.
+    DenseMap<unsigned, unsigned> funcOpResultMapping;
+
+    /// Maintain a mapping between an external memory argument (identified by a
+    /// memref) and eventual component input- and output port indicies that will
+    /// map to the memory ports. The pair denotes the start index of the memory
+    /// ports in the in- and output ports of the component. Ports are expected
+    /// to be ordered in the same manner as they are added by
+    /// getPortsForExternalMemref.
+    DenseMap<Value, std::pair<unsigned, unsigned>> extMemories;
```

Comment:
Can we re-name this to something that indicates the value is indices?

Reply:
will do.